Sources/Overlays/_Testing_Foundation/Attachments/Attachment+URL.swift

```diff
@@ -0,0 +1,161 @@
+private import UniformTypeIdentifiers
+#endif
+
+#if !SWT_NO_FILE_IO
+extension URL {
+  /// The file system path of the URL, equivalent to `path`.
+  var fileSystemPath: String {
+#if os(Windows)
+    // BUG: `path` includes a leading slash which makes it invalid on Windows.
+    // SEE: https://github.com/swiftlang/swift-foundation/pull/964
+    let path = path
+    if path.starts(with: /\/[A-Za-z]:\//) {
+      return String(path.dropFirst())
+    }
+#endif
+    return path
+  }
+}
+
+#if SWT_TARGET_OS_APPLE && canImport(UniformTypeIdentifiers)
+@available(_uttypesAPI, *)
+extension UTType {
+  /// A type that represents a `.tgz` archive, or `nil` if the system does not
+  /// recognize that content type.
+  fileprivate static let tgz = UTType("org.gnu.gnu-zip-tar-archive")
+}
+#endif
+
+@_spi(Experimental)
+extension Attachment where AttachableValue == Data {
+  /// Initialize an instance of this type with the contents of the given URL.
+  ///
+  /// - Parameters:
+  ///   - url: The URL containing the attachment's data.
+  ///   - preferredName: The preferred name of the attachment when writing it to
+  ///     a test report or to disk. If `nil`, the name of the attachment is
+  ///     derived from the last path component of `url`.
+  ///   - sourceLocation: The source location of the call to this initializer.
+  ///     This value is used when recording issues associated with the
+  ///     attachment.
+  ///
+  /// - Throws: Any error that occurs attempting to read from `url`.
+  public init(
+    contentsOf url: URL,
+    named preferredName: String? = nil,
+    sourceLocation: SourceLocation = #_sourceLocation
+  ) async throws {
+    guard url.isFileURL else {
+      // TODO: network URLs?
+      throw CocoaError(.featureUnsupported, userInfo: [NSLocalizedDescriptionKey: "Attaching downloaded files is not supported"])
+    }
+
+    // FIXME: use NSFileCoordinator on Darwin?
+
+    let url = url.resolvingSymlinksInPath()
+    let isDirectory = try url.resourceValues(forKeys: [.isDirectoryKey]).isDirectory!
+
+    // Determine the preferred name of the attachment if one was not provided.
+    var preferredName = if let preferredName {
+      preferredName
+    } else if case let lastPathComponent = url.lastPathComponent, !lastPathComponent.isEmpty {
+      lastPathComponent
+    } else {
+      Self.defaultPreferredName
+    }
+
+    if isDirectory {
+      // Ensure the preferred name of the archive has an appropriate extension.
+      preferredName = {
+#if SWT_TARGET_OS_APPLE && canImport(UniformTypeIdentifiers)
+        if #available(_uttypesAPI, *), let tgz = UTType.tgz {
+          return (preferredName as NSString).appendingPathExtension(for: tgz)
+        }
+#endif
+        return (preferredName as NSString).appendingPathExtension("tgz") ?? preferredName
+      }()
+
+      try await self.init(Data(compressedContentsOfDirectoryAt: url), named: preferredName, sourceLocation: sourceLocation)
```

Comment:
Am I interpreting correctly that for directories, it unconditionally _compresses_ the directory's contents? Could we make the compression step an optional parameter, but enabled by default? (Thinking of scenarios where you're attaching a directory containing files which are already highly compressed.) Maybe disabling that would use `.tar` instead of `.tgz`?

Reply:
This is consistent with what XCTest does here: https://developer.apple.com/documentation/xctest/xctattachment/2887210-init

Reply:
This can be for a future PR, but I'd like to request we revisit that and consider making it configurable. In general, "XCTest did it this way" isn't the most compelling justification 😛